### PR TITLE
Fix $PWD changes when executing Python

### PR DESCRIPTION
**Pull request by zachriggle**
Currently, pyenv uses `cd` in some spots to determine real paths, by abusing `$PWD`.

This works in the general case, but fails when `PWD=/proc/self/cwd`.  When `pyenv` attempts to change directories back to the "old" directory, it fails to do so.

This Pull Request wraps a few of these cases in sub-shells, and changes the code in one instance that uses `$OLDPWD`.

Specifically, this issue becomes apparent because it breaks the build for Android Open Source Project when `pyenv` is on `$PATH`, even when `pyenv global system` is in use.  This is because Clang upstream compilers now use `#!/usr/bin/env python` instead of `#!/usr/bin/python` in their shebang for the Python wrapper scripts.

### CODE REVIEW

File: libexec/pyenv [hunk 1]
```diff
@@ -36,17 +36,18 @@ else
   }
 
   abs_dirname() {
-    local cwd="$PWD"
     local path="$1"
 
+    # Use a subshell to avoid changing the the current path
```
Comment (blueyed) [inline libexec/pyenv]: s/the the/the/

File: libexec/pyenv [hunk 2]
```diff
@@ -72,11 +73,10 @@ fi
 
 if [ -z "${PYENV_DIR}" ]; then
   PYENV_DIR="$PWD"
-else
-  cd "$PYENV_DIR" 2>/dev/null || abort "cannot change working directory to \`$PYENV_DIR'"
-  PYENV_DIR="$PWD"
-  cd "$OLDPWD"
 fi
+
+PYENV_DIR=$(cd "$PYENV_DIR" && echo "$PWD") || abort "cannot change working directory to \`$PYENV_DIR'"
```
Comment (blueyed) [inline libexec/pyenv]: Why is this necessary after all?
Could also check only if it is a directory then?!

Comment (zachriggle) [reply libexec/pyenv]: No idea, I'm trying to respect the current code as closely as possible

**Comment by zachriggle** [comment]
Merging to "upstream" `rbenv` only makes sense if it's possible to merge from `rbenv` to `pyenv` (which I don't think is the case).  I don't use `rbenv` and don't have a good way to test the changes.  I'll open an issue on the `rbenv` issue tracker pointing at this pull request.

File: libexec/pyenv (hunk 1 above)
Comment (zachriggle) [reply libexec/pyenv]: Done

File: libexec/pyenv (hunk 2 above)
Comment (zachriggle) [reply libexec/pyenv]: I've added that code back in to test `-d` and `-x`.  I'm not sure why the check is there in the first place, but there is a test for this specific check.

**Comment by blueyed** [comment]
> if it's possible to merge from rbenv to pyenv (which I don't think is the case)

It is.. I've put some effort into it a while ago, so that it works actually..!

**Comment by yyuu** [comment]
Interesting 👀 

However, I'm still not understanding what is causing the problem in Android Open Source Project.... `pwd` will change inside pyenv before executing `python` itself. Though it should be restored to the original directory before pyenv actually start executing `python`.

Could you provide some minimal script to reproduce the problem...? I've tried writing some, but I could not find issues so far.

```sh
#!/usr/bin/env bash
set -e -o pipefail
export PYENV_ROOT="$(mktemp -d /tmp/pyenv.XXXXXXXX)"
git clone --depth=1 https://github.com/pyenv/pyenv.git "${PYENV_ROOT}"
export PATH="${PYENV_ROOT}/bin:/usr/bin:/bin"
eval "$(pyenv init -)"
# create a dummy version to let pyenv to generate a shim script for python
mkdir -p "${PYENV_ROOT}/versions/foo/bin"
echo -e "#!/bin/sh\necho python \$@" > "${PYENV_ROOT}/versions/foo/bin/python"
chmod 755 "${PYENV_ROOT}/versions/foo/bin/python"
pyenv global system
pwd
cd /proc/self/cwd
python -c 'import os;print(repr(os.getcwd()))'
rm -fr "${PYENV_ROOT}"
```

The `pwd` of `python` process is set as it should be, in my understanding.

```
% ./x.sh
Cloning into '/tmp/pyenv.u1wafgCG'...
remote: Counting objects: 577, done.
remote: Compressing objects: 100% (398/398), done.
remote: Total 577 (delta 268), reused 262 (delta 89), pack-reused 0
Receiving objects: 100% (577/577), 253.72 KiB | 307.00 KiB/s, done.
Resolving deltas: 100% (268/268), done.
/home/yyuu
'/home/yyuu'
```

**Comment by zachriggle** [comment]
The problem is that the `realpath` functions in `pyenv` changes directories before `python` is ever executed:

```
sh -c 'PWD=/proc/self/cwd; cwd="$PWD"; realpath .; cd /; realpath .; cd "$cwd"; realpath .'
# /home/pwntools
# /
# /
```

File: libexec/pyenv [hunk 3]
```diff
@@ -72,11 +73,13 @@ fi
 
 if [ -z "${PYENV_DIR}" ]; then
   PYENV_DIR="$PWD"
-else
-  cd "$PYENV_DIR" 2>/dev/null || abort "cannot change working directory to \`$PYENV_DIR'"
-  PYENV_DIR="$PWD"
-  cd "$OLDPWD"
 fi
+
+if [ ! -d "$PYENV_DIR" ] || [ ! -e "$PYENV_DIR" ]; then
```
Comment (yyuu) [inline libexec/pyenv]: I think just checking with `[ ! -d "${PYENV_DIR}" ]` must be sufficient here.

File: libexec/pyenv [hunk 4]
```diff
@@ -72,11 +73,13 @@ fi
...
+  abort "cannot change working directory to \`$PYENV_DIR'"
+fi
+
+PYENV_DIR=$(cd "$PYENV_DIR" && echo "$PWD")
```
Comment (yyuu) [inline libexec/pyenv]: Just in case, please add quoting around the `$(...)`.

```sh
PYENV_DIR="$(cd "$PYENV_DIR" && echo "$PWD")"
```

File: libexec/pyenv (hunk 2 above)
Comment (blueyed) [reply libexec/pyenv]: > to get the absolute path

This is what I've meant mostly.
At least there does not seem to be a failing test when removing the `PYENV_DIR=$(cd "$PYENV_DIR" && echo "$PWD")`.
Since this spawns a subshell, it is not really cheap, and should be avoided if not necessary.

File: libexec/pyenv (hunk 4 above)
Comment (blueyed) [reply libexec/pyenv]: Might not be necessary altogether: https://github.com/pyenv/pyenv/pull/1036#discussion_r151895707

Comment (yyuu) [reply libexec/pyenv]: I should have followed the discussion little more.... Anyway it's good enough to me to proceed for merging~

**Comment by yyuu** [comment]
Since pyenv is used in various CI provider, I've merged this into pyenv's master to restore the corrupted build of the AOSP.

I or someone who can work for it may be going to port this change to https://github.com/rbenv/rbenv to keep them _compatible_. Because the use of `cd` there in rbenv/pyenv is not anything essential, I don't foresee any issues merging this into rbenv as well.

**Comment by zachriggle** [comment]
Thanks!
